### PR TITLE
Add frame store support

### PR DESCRIPTION
**Pull request by whbruce**
## Build 
```
make vas
<snip>
Successfully tagged video-analytics-serving-rtsf:0.5.0


make cv-roi
<snip>
Successfully tagged rtsf-at-checkout/cv-region-of-interest:dev
```

## Run
```
docker-compose -f loss-detection-app/docker-compose.mqtt.yml -f loss-detection-app/docker-compose.vap.yml up
```
Look for inference results being sent to EdgeX in `enter-exit-event` output
```
Starting mqtt             ... done
Recreating video-analytic ... done
Starting enter-exit-event        ... done
Recreating cv-region-of-interest ... done
Attaching to mqtt, video-analytic, enter-exit-event, cv-region-of-interest
mqtt                     | 1644983682: mosquitto version 1.6.15 starting
mqtt                     | 1644983682: Config loaded from /mosquitto/config/mosquitto.conf.
mqtt                     | 1644983682: Opening ipv4 listen socket on port 1883.
mqtt                     | 1644983682: Opening ipv6 listen socket on port 1883.
mqtt                     | 1644983682: mosquitto version 1.6.15 running
mqtt                     | 1644983683: New connection from 172.18.0.4 on port 1883.
mqtt                     | 1644983683: New client connected from 172.18.0.4 as mosq-Qnm74VeTykksuypOW4 (p2, c1, k60).
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,869", "message": "========================", "module": "vaserving"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,869", "message": "Options for vaserving.py", "module": "vaserving"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,869", "message": "========================", "module": "vaserving"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,869", "message": "port == 8080", "module": "vaserving"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,869", "message": "framework == gstreamer", "module": "vaserving"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,869", "message": "pipeline_dir == pipelines", "module": "vaserving"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,870", "message": "model_dir == models", "module": "vaserving"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,870", "message": "network_preference == {}", "module": "vaserving"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,870", "message": "max_running_pipelines == 1", "module": "vaserving"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,870", "message": "log_level == INFO", "module": "vaserving"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,870", "message": "config_path == /home/video-analytics-serving/vaserving/..", "module": "vaserving"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,870", "message": "ignore_init_errors == False", "module": "vaserving"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,870", "message": "enable_rtsp == True", "module": "vaserving"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,870", "message": "rtsp_port == 8554", "module": "vaserving"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,870", "message": "========================", "module": "vaserving"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,870", "message": "==============", "module": "model_manager"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,870", "message": "Loading Models", "module": "model_manager"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,870", "message": "==============", "module": "model_manager"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,870", "message": "Loading Models from Path /home/video-analytics-serving/models", "module": "model_manager"}
video-analytic           | {"levelname": "WARNING", "asctime": "2022-02-16 03:54:43,870", "message": "Models directory is mount point", "module": "model_manager"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,871", "message": "Loading Model: object_detection version: 1 type: IntelDLDT from {'FP32': '/home/video-analytics-serving/models/object_detection/1/FP32/product-detection-0001.xml', 'model-proc': '/home/video-analytics-serving/models/object_detection/1/grocery.json'}", "module": "model_manager"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,871", "message": "========================", "module": "model_manager"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,871", "message": "Completed Loading Models", "module": "model_manager"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,871", "message": "========================", "module": "model_manager"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,871", "message": "=================", "module": "pipeline_manager"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,871", "message": "Loading Pipelines", "module": "pipeline_manager"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:43,871", "message": "=================", "module": "pipeline_manager"}
mqtt                     | 1644983683: New connection from 172.18.0.5 on port 1883.
mqtt                     | 1644983683: New client connected from 172.18.0.5 as auto-E4592368-866B-9377-9783-1AC763039EDA (p2, c1, k60).
cv-region-of-interest    | Connected to mqtt broker
cv-region-of-interest    | Subscribed to topic
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:44,587", "message": "Loading Pipelines from Config Path /home/video-analytics-serving/pipelines", "module": "pipeline_manager"}
video-analytic           | {"levelname": "WARNING", "asctime": "2022-02-16 03:54:44,588", "message": "Pipelines directory is mount point", "module": "pipeline_manager"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:44,630", "message": "Loading Pipeline: object_detection_cpu version: 2 type: GStreamer from /home/video-analytics-serving/pipelines/object_detection_cpu/2/pipeline.json", "module": "pipeline_manager"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:44,644", "message": "Loading Pipeline: object_detection_cpu version: 1 type: GStreamer from /home/video-analytics-serving/pipelines/object_detection_cpu/1/pipeline.json", "module": "pipeline_manager"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:44,665", "message": "Loading Pipeline: object_detection_cpu version: frame_store type: GStreamer from /home/video-analytics-serving/pipelines/object_detection_cpu/frame_store/pipeline.json", "module": "pipeline_manager"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:44,677", "message": "Loading Pipeline: object_detection version: 2 type: GStreamer from /home/video-analytics-serving/pipelines/object_detection/2/pipeline.json", "module": "pipeline_manager"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:44,691", "message": "Loading Pipeline: object_detection version: 1 type: GStreamer from /home/video-analytics-serving/pipelines/object_detection/1/pipeline.json", "module": "pipeline_manager"}
video-analytic           | {"levelname": "WARNING", "asctime": "2022-02-16 03:54:44,705", "message": "Missing or multiple appsink elements", "module": "gstreamer_pipeline"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:44,705", "message": "Loading Pipeline: object_detection_cpu_render version: 2 type: GStreamer from /home/video-analytics-serving/pipelines/object_detection_cpu_render/2/pipeline.json", "module": "pipeline_manager"}
video-analytic           | {"levelname": "WARNING", "asctime": "2022-02-16 03:54:44,724", "message": "Missing or multiple appsink elements", "module": "gstreamer_pipeline"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:44,724", "message": "Loading Pipeline: object_detection_cpu_render version: 1 type: GStreamer from /home/video-analytics-serving/pipelines/object_detection_cpu_render/1/pipeline.json", "module": "pipeline_manager"}
video-analytic           | {"levelname": "WARNING", "asctime": "2022-02-16 03:54:44,746", "message": "Missing or multiple appsink elements", "module": "gstreamer_pipeline"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:44,747", "message": "Loading Pipeline: object_detection_cpu_restream version: 2 type: GStreamer from /home/video-analytics-serving/pipelines/object_detection_cpu_restream/2/pipeline.json", "module": "pipeline_manager"}
video-analytic           | {"levelname": "WARNING", "asctime": "2022-02-16 03:54:44,766", "message": "Missing or multiple appsink elements", "module": "gstreamer_pipeline"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:44,766", "message": "Loading Pipeline: object_detection_cpu_restream version: 1 type: GStreamer from /home/video-analytics-serving/pipelines/object_detection_cpu_restream/1/pipeline.json", "module": "pipeline_manager"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:44,766", "message": "===========================", "module": "pipeline_manager"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:44,766", "message": "Completed Loading Pipelines", "module": "pipeline_manager"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:44,766", "message": "===========================", "module": "pipeline_manager"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:44,833", "message": "Starting Tornado Server on port: 8080", "module": "__main__"}
cv-region-of-interest    | creating video analytics pipelines
cv-region-of-interest    | Setting up frame store in /tmp/my-frame-store
cv-region-of-interest    | Pipeline endpoint: http://video-analytic:8080/pipelines/object_detection_cpu/frame_store
cv-region-of-interest    | Pipeline request: {"source": {"uri": "file:///tmp/EnterExitEvent_2second.mp4", "type": "uri"}, "destination": {"metadata": {"type": "mqtt", "host": "mqtt:1883", "topic": "AnalyticsData", "timeout": 1000}, "frame": {"type": "rtsp", "path": "cv-roi-event"}}, "tags": {"roi_name": "Staging", "file-location": "/tmp/my-frame-store/img%08d.jpg"}, "parameters": {"top": 0, "left": 0, "right": 0, "bottom": 0, "port": 0, "inference_device": "CPU", "file-location": "/tmp/my-frame-store/img%08d.jpg"}}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:45,006", "message": "Creating Instance of Pipeline object_detection_cpu/frame_store", "module": "pipeline_manager"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:45,009", "message": "Gstreamer RTSP Server Started on port: 8554", "module": "gstreamer_rtsp_server"}
mqtt                     | 1644983685: New connection from 172.18.0.3 on port 1883.
mqtt                     | 1644983685: New client connected from 172.18.0.3 as c052e9e6-d6df-43ed-9bf5-a303869fa813 (p2, c1, k20).
cv-region-of-interest    | Created new pipeline with id: 1
cv-region-of-interest    |
video-analytic           | error: XDG_RUNTIME_DIR not set in the environment.
enter-exit-event         | {"name": "device-cv-roi-mqtt", "cmd": "cv-roi-event", "cv-roi-event": "{\"source\": \"file:///tmp/EnterExitEvent_2second.mp4\", \"event_time\": 1644983686152, \"product_name\": \"Pringles\", \"roi_action\": \"ENTERED\", \"roi_name\": \"Staging\"}"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:46,178", "message": "Created RTSP Stream for instance 1 at rtsp:://<host ip>:8554/cv-roi-event", "module": "gstreamer_rtsp_server"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:46,180", "message": "Setting Pipeline 1 State to RUNNING", "module": "gstreamer_pipeline"}
enter-exit-event         | {"name": "device-cv-roi-mqtt", "cmd": "cv-roi-event", "cv-roi-event": "{\"source\": \"file:///tmp/EnterExitEvent_2second.mp4\", \"event_time\": 1644983688426, \"product_name\": \"Sprite 2L\", \"roi_action\": \"ENTERED\", \"roi_name\": \"Staging\", \"frame_path\": \"/tmp/my-frame-store/img00000060.jpg\"}"}
enter-exit-event         | {"name": "device-cv-roi-mqtt", "cmd": "cv-roi-event", "cv-roi-event": "{\"source\": \"file:///tmp/EnterExitEvent_2second.mp4\", \"event_time\": 1644983688426, \"product_name\": \"Ruffles\", \"roi_action\": \"ENTERED\", \"roi_name\": \"Staging\", \"frame_path\": \"/tmp/my-frame-store/img00000060.jpg\"}"}
enter-exit-event         | {"name": "device-cv-roi-mqtt", "cmd": "cv-roi-event", "cv-roi-event": "{\"source\": \"file:///tmp/EnterExitEvent_2second.mp4\", \"event_time\": 1644983690099, \"product_name\": \"Sprite 2L\", \"roi_action\": \"ENTERED\", \"roi_name\": \"Staging\", \"frame_path\": \"/tmp/my-frame-store/img00000120.jpg\"}"}
enter-exit-event         | {"name": "device-cv-roi-mqtt", "cmd": "cv-roi-event", "cv-roi-event": "{\"source\": \"file:///tmp/EnterExitEvent_2second.mp4\", \"event_time\": 1644983690099, \"product_name\": \"Pringles\", \"roi_action\": \"EXITED\", \"roi_name\": \"Staging\", \"frame_path\": \"/tmp/my-frame-store/img00000120.jpg\"}"}
enter-exit-event         | {"name": "device-cv-roi-mqtt", "cmd": "cv-roi-event", "cv-roi-event": "{\"source\": \"file:///tmp/EnterExitEvent_2second.mp4\", \"event_time\": 1644983691790, \"product_name\": \"Sprite 2L\", \"roi_action\": \"EXITED\", \"roi_name\": \"Staging\", \"frame_path\": \"/tmp/my-frame-store/img00000180.jpg\"}"}
enter-exit-event         | {"name": "device-cv-roi-mqtt", "cmd": "cv-roi-event", "cv-roi-event": "{\"source\": \"file:///tmp/EnterExitEvent_2second.mp4\", \"event_time\": 1644983691790, \"product_name\": \"Sprite 2L\", \"roi_action\": \"EXITED\", \"roi_name\": \"Staging\", \"frame_path\": \"/tmp/my-frame-store/img00000180.jpg\"}"}
enter-exit-event         | {"name": "device-cv-roi-mqtt", "cmd": "cv-roi-event", "cv-roi-event": "{\"source\": \"file:///tmp/EnterExitEvent_2second.mp4\", \"event_time\": 1644983691790, \"product_name\": \"Ruffles\", \"roi_action\": \"EXITED\", \"roi_name\": \"Staging\", \"frame_path\": \"/tmp/my-frame-store/img00000180.jpg\"}"}
video-analytic           | {"levelname": "INFO", "asctime": "2022-02-16 03:54:56,776", "message": "Pipeline 1 Ended", "module": "gstreamer_pipeline"}
mqtt                     | 1644983696: Client c052e9e6-d6df-43ed-9bf5-a303869fa813 disconnected.
```

## PR details
* Frame store-enabled pipeline is called object_detection_cpu/frame_store. It uses a plug-in that’s in the loss-detection-app/extensions folder.
* docker-compose.vap.yml has a number of updates
  * sets user and mounts /tmp folder for VA Serving and enter/exit service containers so they can both access frame-store
  * mounts extension folders so VA Serving has access to pipeline's gvapython plug-in
  * selects frame_store enabled pipeline
  * defines location of the frame store via CAMERA0_FRAME_STORE environment variable. 
  * includes an mqtt broker and subscriber to show the enter/exit events coming through.
* The enter-exit service creates the frame store directory and ensure appropriate permissions.
* Path to frame is in the “frame_path” property of the enter/exit edgex message. See example below where frames is at `/tmp/my-frame-store/img00000060.jpg`
```json
{
   "name": "device-cv-roi-mqtt", 
   "cmd": "cv-roi-event", 
   "cv-roi-event": "{\"source\": \"file:///tmp/EnterExitEvent_2second.mp4\", \"event_time\": 1620941451885, \"product_name\": \"Sprite 2L\", \"roi_action\": \"ENTERED\", \"roi_name\": \"Staging\", \"frame_path\": \"/tmp/my-frame-store/img00000060.jpg\"}"
}

### CODE REVIEW

File: rtsf-at-checkout-cv-region-of-interest/src/enterexit.py [hunk 1]
```diff
@@ -15,33 +17,49 @@
...
 EDGEX_ENTER_EVENT = 'ENTERED'
 EDGEX_EXIT_EVENT = 'EXITED'
-
 MQTT_BROKER_ADDRESS = MQTT_BROKER_HOST + ":" + str(MQTT_BROKER_PORT)
 
+FRAME_STORE_SPECIFIER = "img%08d.jpg"
+
 oldFrameDict = {}
 
+def create_event_message(source, key, event_type, roi_name, frame_path):
```
Comment (brian-intel) [inline rtsf-at-checkout-cv-region-of-interest/src/enterexit.py:26]: Great refactor

**Review by brian-intel — APPROVED** [review]

File: loss-detection-app/docker-compose.vap.yml [hunk 2]
```diff
@@ -2,44 +2,70 @@ version: '3.4'
...
+    image: eclipse-mosquitto:1.6
+    container_name: enter-exit-event
+    hostname: enter-exit-event
+    networks:
+      - loss-detection-app_edgex-network
+    entrypoint: /usr/bin/mosquitto_sub -h mqtt -t edgex
+    depends_on:
+      - mqtt
+
+  mqtt:
```
[the remark this answers is not in the document]
Comment (whbruce) [reply loss-detection-app/docker-compose.vap.yml]: I moved vap test mqtt services into their own compose file so I can now test like this
```
docker-compose -f loss-detection-app/docker-compose.mqtt.yml -f loss-detection-app/docker-compose.vap.yml up
```
@brian-intel I'll leave you to resolve the converation.

File: rtsf-at-checkout-cv-region-of-interest/src/enterexit.py (hunk 1 above)
Comment (whbruce) [reply rtsf-at-checkout-cv-region-of-interest/src/enterexit.py:26]: Thx!

**Comment by jim-wang-intel** [comment]
this one is out of date, closing it

**Pull request closed without merging**